eodatasets3/scripts/tostac.py

```diff
@@ -2,75 +2,214 @@
-            properties={**dataset.properties, "odc:product": dataset.product.name},
-            assets={
-                # TODO: Currently assuming no name collisions.
-                **{name: {"href": m.path} for name, m in dataset.measurements.items()},
-                **{name: {"href": m.path} for name, m in dataset.accessories.items()},
-            },
-            links=[
-                # {
-                #     "rel": "self",
-                #     "href": '?',
-                # },
-                {"rel": "odc_product", "href": dataset.product.href},
-                {
-                    "rel": "alternative",
-                    "type": "text/html",
-                    "href": f"https://explorer.dea.ga.gov.au/dataset/{dataset.id}",
-                },
-            ],
-        )
-
         name = input_metadata.stem.replace(".odc-metadata", "")
         output_path = input_metadata.with_name(f"{name}.stac-item.json")
 
+        # Load STAC template data
+        if stac_template:
+            with open(stac_template, "r") as f:
+                stac_data = json.load(f)
+                if stac_data.get("stac_version", "") not in [
+                    "1.0.0-beta.1",
+                    "1.0.0-beta.2",
+                ]:
+                    raise NotImplementedError(
+                        f"STAC version {stac_data.get('stac_version', '')} not implemented"
+                    )
+        else:
+            stac_data = {}
+
+        # Create STAC dict
+        item_doc = dc_to_stac(
+            dataset,
+            input_metadata,
+            output_path,
+            stac_data,
+            stac_base_url,
+            explorer_base_url,
+            validate,
+        )
+
         with output_path.open("w") as f:
             json.dump(item_doc, f, indent=4, default=json_fallback)
 
-        echo(output_path)
+
+def dc_to_stac(
+    dataset: DatasetDoc,
+    input_metadata: Path,
+    output_path: Path,
+    stac_data: dict,
+    stac_base_url: str,
+    explorer_base_url: str,
+    do_validate: bool,
+) -> dict:
+    """
+    Creates a STAC document
+    """
+
+    stac_ext = ["eo", "view", "projection"]
+    stac_ext.extend(stac_data.get("stac_extensions", []))
+
+    geom = Geometry(dataset.geometry, CRS(dataset.crs))
+    wgs84_geometry = geom.to_crs(CRS("epsg:4326"), math.inf)
+
+    item_doc = dict(
+        stac_version="1.0.0-beta.2",
+        stac_extensions=sorted(set(stac_ext)),
+        type="Feature",
+        id=dataset.id,
+        bbox=wgs84_geometry.boundingbox,
+        geometry=wgs84_geometry.json,
+        properties={
+            **{
+                MAPPING_EO3_TO_STAC.get(key, key): convert_value_to_stac_type(key, val)
+                for key, val in dataset.properties.items()
+            },
+            "odc:product": dataset.product.name,
+            "proj:epsg": int(dataset.crs.lstrip("epsg:")) if dataset.crs else None,
+            "proj:shape": dataset.grids["default"].shape,
```

Comment:
A 'default' grid is not guaranteed to be there... it's just the name used when they don't specify a grid name. The least-surprising behaviour would be to pick the most frequently-used grid (since I usually see all bands bar-one use the same grid)

(though I haven't looked into how people usually treat datasets with multple grids with this extension)

Reply:
Check added if no `default` grid, fields `proj:shape` and `proj:transform` are not added.